invertedai_cpp/invertedai/blame_response.cc

```diff
@@ -0,0 +1,105 @@
+#include "blame_response.h"
+
+using json = nlohmann::json;
+
+namespace invertedai {
+
+BlameResponse::BlameResponse(const std::string &body_str) {
+  this->body_json_ = json::parse(body_str);
+
+  this->agents_at_fault_.clear();
+  for (const auto &element : this->body_json_["agents_at_fault"]) {
+    this->agents_at_fault_.push_back(element.get<int>());
+  }
+  this->confidence_score_ =
+      this->body_json_["confidence_score"].is_number()
+          ? std::optional<float>{this->body_json_["confidence_score"]
+                                     .get<float>()}
+          : std::nullopt;
+  if (this->body_json_["reasons"].is_null()) {
+    this->reasons_ = std::nullopt;
+  } else {
+    std::map<int, std::vector<std::string>> reasons;
+    reasons.clear();
+    for (const auto &pair :
+         this->body_json_["reasons"]
+             .get<std::map<std::string, std::vector<std::string>>>()) {
+      reasons.insert(std::make_pair(std::stoi(pair.first), pair.second));
+    }
+    this->reasons_ =
+        std::optional<std::map<int, std::vector<std::string>>>{reasons};
+  }
+  if (this->body_json_["birdviews"].is_null()) {
+    this->birdviews_ = std::nullopt;
+  } else {
+    this->birdviews_ = std::optional<std::vector<std::vector<unsigned char>>>{
+        this->body_json_["birdviews"]
+            .get<std::vector<std::vector<unsigned char>>>()};
+  }
+}
+
+void BlameResponse::refresh_body_json_() {
+  this->body_json_["agents_at_fault"] = this->agents_at_fault_;
+  if (this->confidence_score_.has_value()) {
+    this->body_json_["confidence_score"] = this->confidence_score_.value();
+  } else {
+    this->body_json_["confidence_score"] = nullptr;
+  }
+  if (this->reasons_.has_value()) {
+    this->body_json_["reasons"] = json::object();
+    for (const auto &pair : this->reasons_.value()) {
+      this->body_json_["reasons"][std::to_string(pair.first)] = pair.second;
+    }
+  } else {
+    this->body_json_["reasons"] = nullptr;
+  }
+  if (this->birdviews_.has_value()) {
+    this->body_json_["birdviews"] = this->birdviews_.value();
+  } else {
+    this->body_json_["birdviews"] = nullptr;
+  }
+}
+
+std::string BlameResponse::body_str() {
+  this->refresh_body_json_();
+  return this->body_json_.dump();
+}
+
+std::vector<int> BlameResponse::agents_at_fault() const {
+  return this->agents_at_fault_;
+}
```

Comment:
Just to confirm, agents_at_fault is NOT implemented as the type std::optional<...> because there is no flag option to turn off this value like reasons, confidence_score, and birdviews, correct? If that is the reason I'm wondering why we don't implement std::optional<...> for similar parameters in our other C++ response endpoints. Perhaps this is a matter for another PR if std::optional<...> is better.

Reply:
Yeah, currently the `optional` feature in the C++ responses is not strictly taken care of. I tried the REST api and found that even though when the `get_***` flags are turned off, the REST responses will contain default values for the corresponding fields instead of returning null, and thus won't cause issues to the C++ response parsing. I agree that we can wrap the fields more clearly. I will create another PR to address it.